### PR TITLE
[pkg/ottl] Add Log() to ottlfuncs

### PR DESCRIPTION
**Pull request by mterhar**
Addresses issue https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/18076

Description:

Adds a feature to allow OTTL expressions to get a logarithm of a value.

Link to tracking Issue: 18076

Testing: Tests were added in the same way the Int() function is tested

Documentation: Added a readme section following the same pattern as Int() which covers type for inputs and return values and examples.

### CODE REVIEW

File: pkg/ottl/ottlfuncs/func_log.go [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+			fltValue, err := strconv.ParseFloat(value, 64)
+			if err != nil {
+				return nil, nil
+			}
+
+			return math.Log(fltValue), nil
+		case float64:
+			return math.Log(value), nil
+		default:
+			return nil, nil
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_log.go:64]: We have a new concept of typed-getters that we should use here instead:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/586c62e2241cd18e77a47cd894766a3bdb01376a/pkg/ottl/expression.go#L146-L226

In this case it looks like we want a float64, so we could add a new `FloatLikeGetter` which would behave similar to `StringLikeGetter`.  I'd also be ok with this function being stricter and only allowing int64/float64 and not handling string conversion, in which case we could add a `Float64Getter` that would allow floats and ints.

You can do that as part of this PR or in a separate PR, or if you want I can add it in a separate PR.

Comment (mterhar) [reply pkg/ottl/ottlfuncs/func_log.go:64]: I'm going to leave this out of the first cut since I'm not sure people using untyped languages are going to be sure that 2.3 is different from "2.3" and don't want to throw unsupported type error when we can try to parse it.

File: pkg/ottl/ottlfuncs/func_log_test.go [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+	"context"
+	"math"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+func Test_Log(t *testing.T) {
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/func_log_test.go:27]: Please add tests for `0` and negative number.

File: pkg/ottl/ottlfuncs/func_log.go (hunk 1 above)
Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_log.go]: I need to document it better, but with the improvements we've made to OTTL's error handling it is good for functions to return errors when they encounter unexpected or undesired situations.  By using typed-getters and returning errors we create consistency across all the OTTL functions.  If users want to ignore the errors they can use `error_mode=ignore`, which is the recommended error mode.

Unfortunately I haven't gotten around to making all the typed-getters so we can still get into situations like these where we need a new one.  I opened up https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/21896 to add Float Getters that this function will use.  If you don't want to wait for that PR feel free to add those changes to this PR and use `FloatLikeGetter` and I'll close mine.

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_log.go]: Actually, lets wait for that PR to merge then do a rebase.  It should get merged today.  It was a large lift than I anticipated and I want to keep the commit history clean now.

Comment (TylerHelmuth) [reply pkg/ottl/ottlfuncs/func_log.go]: @mterhar please get the latest from main and use `FloatLikeGetter` as the Getter type.

Comment (mterhar) [reply pkg/ottl/ottlfuncs/func_log.go]: Let me know if I missed something about the update but it looks like it's working with the better getter.

File: pkg/ottl/ottlfuncs/README.md [hunk 3]
```diff
@@ -364,6 +365,29 @@ Examples:
...
+
+The returned type is float64.
+
+The input `value` types:
+* float64, string, and Int64. Converts to a Float64 and returns `math.Log()` of the converted value
+* bool. Returns `nil` because log(1) is zero and log(0) is undefined so neither are useful.
+
+If `value` is another type or parsing failed nil is always returned.
+
+The `value` is either a path expression to a telemetry field to retrieve or a literal.
```
Comment (TylerHelmuth) [inline pkg/ottl/ottlfuncs/README.md:382]:
```suggestion
The `Log` Converter returns the logarithm of the `target`.

`target` is either a path expression to a telemetry field to retrieve or a literal.

The function take the logarithm of the target, returning an error if the target is less than or equal to zero.

If target is not a float64, it will be converted to one:

- int64s are converted to float64s
- strings are converted using `strconv`
- booleans are converted using `1` for `true` and `0` for `false`.  This means passing `false` to the function will cause an error.
- int, float, string, and bool OTLP Values are converted following the above rules depending on their type.  Other types cause an error.

If target is nil an error is returned.
```

**Review by TylerHelmuth — APPROVED** [review]
Left a suggestion for the README to bring it in line with `IsMatch`, which is the only other function using a `Like` getter.  Otherwise LGTM
